cep-20.1.md

```diff
@@ -0,0 +1,286 @@
+<tr><td> Status </td><td> Proposed</td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
+<tr><td> Created </td><td> May 23, 2023</td></tr>
+<tr><td> Updated </td><td> May 23, 2023</td></tr>
+<tr><td> Discussion </td><td>  </td></tr>
+<tr><td> Implementation </td><td>https://github.com/prefix-dev/rattler-build</td></tr>
+</table>
+
+## Abstract
+
+We propose a new recipe format that is heavily inspired by conda-build. The main
+change is a pure YAML format without arbitrary Jinja or comments with semantic
+meaning.
+
+## Motivation
+
+The conda-build format has grown over the years to become quite complex.
+Unfortunately it has never been formally "specified" and it has grown some
+features over time that make it hard to parse as straightforward YAML.
+
+The CEP attempts to introduce a subset of the conda build format that allows for
+fast parsing and building of recipes.
+
+### History
+
+A discussion was started on what a new recipe spec could or should look like.
+The fragments of this discussion can be found here:
+https://github.com/mamba-org/conda-specs/blob/master/proposed_specs/recipe.md
+The reason for a new spec are:
+
+- Make it easier to parse ("pure yaml"). conda-build uses a mix of comments and
+  jinja to achieve a great deal of flexibility, but it's hard to parse the
+  recipe with a computer
+- iron out some inconsistencies around multiple outputs (build vs. build/script
+  and more)
+- remove any need for recursive parsing & solving
+
+## Major differences with conda-build
+
+- no full Jinja2 support: no conditional or `{% set ...` support, only string
+  interpolation. Variables can be set in the toplevel "context" which is valid
+  YAML
+- Jinja string interpolation needs to be quoted at the beginning of a string,
+  e.g. `- "{{ version }}"` in order for it to be valid YAML
+- Selectors use a YAML dictionary style (vs. comments in conda-build). E.g. `-
+  sel(osx): somepkg` instead of `- somepkg  # [osx]`
+
+## Selectors
+
+Selectors in the new spec take the following format:
+
+`sel(unix): selected_value`
+
+This is a valid YAML dictionary. Selector contents are simple boolean
+expressions and follow Python syntax. The following selectors are all valid:
+
+```
+win and arm64
+(osx or linux) and aarch64
+something == "test"
+```
+
+### The cmp function for variant selection
+
+Furthermore, we have a special "cmp" function that can be used to run a check
+against a selected variant version. The `cmp` function looks like the following:
+
+```
+cmp(python, "3.6")
+cmp(python, ">=3.6")
+cmp(python, ">=3.8,<3.10")
+etc
+```
+
+This can be used in a selector like so:
+
+```
+requirements:
+  build:
+    - sel(cmp(python, ">=3.6,<3.10")): dataclasses
+```
+
+This functionality generalizes and replaces the previous special variables such
+as `py2k`, `py3k`, `py36`, `py37`, and works just as well for NumPy, Ruby, R, or
+any other variant that might be of interest in the future.
```

Comment:
One follow up question, can we assign these somehow?

Maybe users could roll their own `py36` or similar. This might help with readability when there are a lot of these

```yaml
context:
  py36 = cmp(python, ">=3.6,<3.10")
```

Another thought could users have multiple keys under a selector. For example

```yaml
  build:
    - sel(cmp(python, ">=3.6,<3.10")):
      - dataclasses
      - importlib_metadata
```

Maybe both together?


```yaml
context:
  py36 = cmp(python, ">=3.6,<3.10")

...

  build:
    - py36:
      - dataclasses
      - importlib_metadata
```

Reply:
I haven't thought about assigning a boolean variable in the context but it should indeed be possible. 

Regarding the list-merging, that is indeed mentioned in the CEP and what we want to implement.